### PR TITLE
[Enhancement] Delete garbage data throught recycle bin after truncate table in share data mode

### PR DESCRIPTION
**Pull request by srlch**
Using recycle bin to delete file after truncate table which is more faster than using StarMgrMetaSyncer

## What I'm doing:

Fixes #issue

## What type of PR is this:

- [ ] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function
- [ ] This is a backport pr

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.3
  - [ ] 3.2
  - [ ] 3.1
  - [ ] 3.0
  - [ ] 2.5

### CODE REVIEW

File: fe/fe-core/src/test/java/com/starrocks/catalog/ReplaceLakePartitionTest.java [hunk 1]
```diff
@@ -91,11 +112,153 @@ LakeTable buildLakeTableWithTempPartition(PartitionType partitionType) {
...
+            @Mock
+            public Warehouse getBackgroundWarehouse() {
+                return new DefaultWarehouse(WarehouseManager.DEFAULT_WAREHOUSE_ID, WarehouseManager.DEFAULT_WAREHOUSE_NAME);
+            }
+        };
+
+        try {
+            GlobalStateMgr.getCurrentState().getRecycleBin().erasePartition(Long.MAX_VALUE);
+        } catch (Exception ignore) {
+        }
```
Comment (kevincai) [inline fe/fe-core/src/test/java/com/starrocks/catalog/ReplaceLakePartitionTest.java:166]:
```suggestion
        ExceptionChecker.expectThrowsNoException(() -> GlobalStateMgr.getCurrentState().getRecycleBin().erasePartition(Long.MAX_VALUE);
```

File: fe/fe-core/src/main/java/com/starrocks/catalog/CatalogRecycleBin.java [hunk 2]
```diff
@@ -902,6 +902,11 @@ protected void runAfterCatalogReady() {
         }
     }
 
+    @VisibleForTesting
+    public synchronized boolean isContainedInidToRecycleTime(long id) {
```
Comment (kevincai) [inline fe/fe-core/src/main/java/com/starrocks/catalog/CatalogRecycleBin.java:906]:
```suggestion
    synchronized boolean isContainedInidToRecycleTime(long id) {
```

remove the access identifier, so the interface can be accessible from the same package, but not visible outside the package, which can be a workaround to allow UT accessing.

File: fe/fe-core/src/main/java/com/starrocks/catalog/OlapTable.java [hunk 3]
```diff
@@ -2105,6 +2105,15 @@ protected OlapTable selectiveCopyInternal(OlapTable copied, Collection<String> r
     public Partition replacePartition(Partition newPartition) {
         Partition oldPartition = nameToPartition.remove(newPartition.getName());
 
+        // For cloud native table, add partition into recycle Bin after truncate table.
+        // It is no necessary for share nothing mode because file will be deleted throught
+        // tablet report in this case.
+        if (this.isCloudNativeTable()) {
```
Comment (kevincai) [inline fe/fe-core/src/main/java/com/starrocks/catalog/OlapTable.java:2111]: shall be applicable to shared-data materialized view as well?

File: fe/fe-core/src/test/java/com/starrocks/catalog/ReplaceLakePartitionTest.java [hunk 4]
```diff
@@ -91,11 +112,153 @@ LakeTable buildLakeTableWithTempPartition(PartitionType partitionType) {
...
+                return shardInfo;
+            }
+        };
+
+        new MockUp<WarehouseManager>() {
+            @Mock
+            public Warehouse getBackgroundWarehouse() {
+                return new DefaultWarehouse(WarehouseManager.DEFAULT_WAREHOUSE_ID, WarehouseManager.DEFAULT_WAREHOUSE_NAME);
+            }
+        };
```
Comment (kevincai) [inline fe/fe-core/src/test/java/com/starrocks/catalog/ReplaceLakePartitionTest.java:204]: check if possible set these common expectations in a function and call the function to set MockUp/Expectations repeatedly.

File: fe/fe-core/src/main/java/com/starrocks/catalog/CatalogRecycleBin.java (hunk 2 above)
Comment (srlch) [reply fe/fe-core/src/main/java/com/starrocks/catalog/CatalogRecycleBin.java]: Updated

File: fe/fe-core/src/main/java/com/starrocks/catalog/OlapTable.java (hunk 3 above)
Comment (srlch) [reply fe/fe-core/src/main/java/com/starrocks/catalog/OlapTable.java]: Updated

File: fe/fe-core/src/test/java/com/starrocks/catalog/ReplaceLakePartitionTest.java (hunk 4 above)
Comment (srlch) [reply fe/fe-core/src/test/java/com/starrocks/catalog/ReplaceLakePartitionTest.java]: Updated

**Review by kevincai — APPROVED** [review]

**Review by HangyuanLiu — APPROVED** [review]

**Pull request merged**